src/broadcast.jl

```diff
@@ -0,0 +1,147 @@
+                newsize[j] = s[j]
+            else
+                throw(DimensionMismatch("Tried to broadcast on inputs sized $sizes"))
+            end
+        end
+    end
+    newsize = tuple(newsize...)
+
+    exprs = Array{Expr}(newsize)
+    more = true
+    current_ind = ones(Int, length(newsize))
+
+    while more
+        exprs_vals = [(a[i] <: Number ? :(a[$i]) : :(a[$i][$(broadcasted_index(sizes[i], current_ind))])) for i = 1:length(sizes)]
+        exprs[current_ind...] = :(f($(exprs_vals...)))
+
+        # increment current_ind
+        current_ind[1] += 1
+        for i ∈ 1:length(newsize)
+            if current_ind[i] > newsize[i]
+                if i == length(newsize)
+                    more = false
+                    break
+                else
+                    current_ind[i] = 1
+                    current_ind[i+1] += 1
+                end
+            else
+                break
+            end
+        end
+    end
+
+    eltype_exprs = [:(eltype($t)) for t ∈ a]
+    newtype_expr = :(Core.Inference.return_type(f, Tuple{$(eltype_exprs...)}))
+
+    return quote
+        @_inline_meta
+        @inbounds return similar_type($first_staticarray, $newtype_expr, Size($newsize))(tuple($(exprs...)))
+    end
+end
+
+
+################
+## broadcast! ##
+################
+
+@propagate_inbounds function broadcast!(f, dest::StaticArray, a::Union{Number, StaticArray}...)
+    _broadcast!(f, Size(dest), dest, broadcast_sizes(a...), a...)
+end
+
+
+@generated function _broadcast!(f, ::Size{newsize}, dest::StaticArray, s::Tuple{Vararg{Size}}, a::Union{Number, StaticArray}...) where {newsize}
+    sizes = [sz.parameters[1] for sz ∈ s.parameters]
+    sizes = tuple(sizes...)
+
+    ndims = 0
+    for i = 1:length(sizes)
+        ndims = max(ndims, length(sizes[i]))
+    end
+
+    for i = 1:length(sizes)
+        s = sizes[i]
+        for j = 1:length(s)
+            if s[j] != 1 && s[j] != (j <= length(newsize) ? newsize[j] : 1)
+                throw(DimensionMismatch("Tried to broadcast to destination sized $newsize from inputs sized $sizes"))
+            end
+        end
+    end
+
+    exprs = Array{Expr}(newsize)
+    j = 1
+    more = true
+    current_ind = ones(Int, max(length(newsize), length.(sizes)...))
+    while more
+        exprs_vals = [(a[i] <: Number ? :(a[$i]) : :(a[$i][$(broadcasted_index(sizes[i], current_ind))])) for i = 1:length(sizes)]
+        exprs[current_ind...] = :(dest[$j] = f($(exprs_vals...)))
+
+        # increment current_ind
+        current_ind[1] += 1
+        for i ∈ 1:length(newsize)
+            if current_ind[i] > newsize[i]
+                if i == length(newsize)
+                    more = false
+                    break
+                else
+                    current_ind[i] = 1
+                    current_ind[i+1] += 1
+                end
+            else
+                break
+            end
+        end
+        j += 1
+    end
```

Comment:
This loop looks exactly the same as in `broadcast`, with the exception of the `dest[$j] =` bit.  Could be factored together?